Fix suppressions for librdkafka data-race for statistics code

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix suppressions for librdkafka data-race for statistics code

The problem is that ignorelist `fun` does not work recursively.

<details>

<summary>example</summary>

```c
// test.cpp

#include <thread>

bool flag = false;

// avoid mangling
extern "C" {

void set_flag_impl()
{
    flag = true;
}
void set_flag()
{
    set_flag_impl();
}
void set_flag_if()
{
    if (flag)
        flag = false;
}

}

int main()
{
    std::thread t1([]{ set_flag(); });
    std::thread t2([]{ set_flag_if(); });

    t1.join();
    t2.join();

    return 0;
}
```

```
// ignorelist
[thread]
fun:set_flag
```

```
$ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
SUMMARY: ThreadSanitizer: data race /tmp/tsan-test/test.cpp:18:9 in set_flag_if
$ sed -i 's/set_flag/set_flag_impl/' ignorelist
$ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
OK
```

</details>

Fixes: https://github.com/ClickHouse/ClickHouse/issues/60939 (cc @Algunenano )
Fixes: https://github.com/ClickHouse/ClickHouse/issues/60443
Fixes: https://github.com/ClickHouse/ClickHouse/pull/60604
Refs: https://github.com/ClickHouse/ClickHouse/pull/61828

azat committed:
Remove outdated comment about broken -fsanitize-ignorelist

See: https://github.com/ClickHouse/ClickHouse/pull/49829#issuecomment-1546704055
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Use sanitizer specific ignorelists

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix suppressions for librdkafka data-race for statistics code

The problem is that ignorelist `fun` does not work recursively.

<details>

<summary>example</summary>

```c
// test.cpp

bool flag = false;

// avoid mangling
extern "C" {

void set_flag_impl()
{
    flag = true;
}
void set_flag()
{
    set_flag_impl();
}
void set_flag_if()
{
    if (flag)
        flag = false;
}

}

int main()
{
    std::thread t1([]{ set_flag(); });
    std::thread t2([]{ set_flag_if(); });

    t1.join();
    t2.join();

    return 0;
}
```

```
// ignorelist
[thread]
fun:set_flag
```

```
$ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
SUMMARY: ThreadSanitizer: data race /tmp/tsan-test/test.cpp:18:9 in set_flag_if
$ sed -i 's/set_flag/set_flag_impl/' ignorelist
$ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
OK
```

</details>

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #63039 from azat/ClickHouse:tsan-suppressions-fix -->
*This is an automated comment for commit 84bebc534ba7cf6e9dbfc1d91e8350939a84f87c with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63039/84bebc534ba7cf6e9dbfc1d91e8350939a84f87c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

azat committed:
Fix suppressions for rd_avg_calc()/rd_avg_rollover() (due to static qualifier)

The problem is that TSan still fails [1] is that ignorelist does not
work for static functions without asterisk:

    // test.cpp
    #include <thread>

    bool flag = false;

    // avoid mangling
    extern "C" {

    static void set_flag_impl()
    {
        flag = true;
    }

    void set_flag()
    {
        set_flag_impl();
    }
    void set_flag_if()
    {
        if (flag)
            flag = false;
    }

    }

    int main()
    {
        std::thread t1([]{ set_flag(); });
        std::thread t2([]{ set_flag_if(); });

        t1.join();
        t2.join();

        return 0;
    }

    // ignorelist
    [thread]
    fun:set_flag_impl

    $ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
    SUMMARY: ThreadSanitizer: data race /tmp/test-tsan-ignorelist/test.cpp:19:9 in set_flag_if
    $ sed -i 's/set_flag_impl/*set_flag_impl*/' ignorelist
    $ clang++ -g -fno-omit-frame-pointer -fsanitize=thread -fsanitize-ignorelist=ignorelist -o test test.cpp && ./test
    OK

But, note that ignorelist is tricky, and will not work for
functions with __always_inline__ attribute for example.

P.S. set_flag_impl also has brackets in the output (i.e.
set_flag_impl()), while ther eis brackets for rd_avg_calc on CI [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/63039/84bebc534ba7cf6e9dbfc1d91e8350939a84f87c/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

antaljanosbenjamin commented:
@azat it seems the ignore list doesn't work, or at least that's what I concluded as this popped up [again](https://s3.amazonaws.com/clickhouse-test-reports/63807/98ad4844b0f6f8ab79a3da1511eb0f2707550faf/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log) (in May). I checked the llvm docs, your PR, etc and it seems fine. Do you have any idea what could go wrong?
